File: Go/Day-14/problem-1/problem-1-solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	RoomWidth   = 101
	RoomHeight  = 103
	TimeElapsed = 100
)

type Robot struct {
	x, y   int // Position as grid coordinates
	vx, vy int // Velocity in tiles per second
}

func (r *Robot) UpdatePosition() {
	// Move discretely for 100 seconds
	r.x = discreteWrap(r.x+r.vx*TimeElapsed, RoomWidth)
	r.y = discreteWrap(r.y+r.vy*TimeElapsed, RoomHeight)
}

// Discrete wrapping function for grid-based movement
func discreteWrap(pos, size int) int {
	// Handle wrapping for both positive and negative velocities
	for pos < 0 {
		pos += size
	}
	return pos % size
}

func determineQuadrant(x, y int) int {
	// Exclude middle lines exactly
	if x == RoomWidth/2 || y == RoomHeight/2 {
		return -1 // Not in any quadrant
	}

	// Quadrant numbering:
	// 0 | 1
	// -----
	// 2 | 3
	if x < RoomWidth/2 && y < RoomHeight/2 {
		return 0
	} else if x > RoomWidth/2 && y < RoomHeight/2 {
		return 1
	} else if x < RoomWidth/2 && y > RoomHeight/2 {
		return 2
	} else {
		return 3
	}
}

func main() {
	// Read input from file
	file, err := os.Open("inputdata.txt")
	if err != nil {
		log.Fatal("Error opening input file:", err)
	}
	defer file.Close()

	var robots []Robot
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, " ")

		// Parse position
		pParts := strings.Split(strings.TrimPrefix(parts[0], "p="), ",")
		x, _ := strconv.Atoi(pParts[0])
		y, _ := strconv.Atoi(pParts[1])

		// Parse velocity
		vParts := strings.Split(strings.TrimPrefix(parts[1], "v="), ",")
		vx, _ := strconv.Atoi(vParts[0])
		vy, _ := strconv.Atoi(vParts[1])

		robots = append(robots, Robot{
			x:  x,
			y:  y,
			vx: vx,
			vy: vy,
		})
	}

	// Debug: Show initial robot positions and movements
	fmt.Println("Initial Robots:")
	for i, robot := range robots {
		fmt.Printf("Robot %d: Initial(x=%d,y=%d) Velocity(vx=%d,vy=%d)\n",
			i, robot.x, robot.y, robot.vx, robot.vy)
	}

	// Update robot positions
	for i := range robots {
		robots[i].UpdatePosition()
	}

	// Debug: Show final robot positions
	fmt.Println("\nFinal Robots:")
	for i, robot := range robots {
		fmt.Printf("Robot %d: Final(x=%d,y=%d)\n", i, robot.x, robot.y)
	}

	// Count robots in quadrants
	quadrantCounts := make([]int, 4)
	for _, robot := range robots {
		quadrant := determineQuadrant(robot.x, robot.y)
		if quadrant >= 0 {
			quadrantCounts[quadrant]++
		}
	}

	// Calculate safety factor
	safetyFactor := 1
	for _, count := range quadrantCounts {
		safetyFactor *= count
	}

	// Print results
	fmt.Println("\nQuadrant Counts:", quadrantCounts)
	fmt.Println("Safety Factor:", safetyFactor)
}

File: 2024/C/14/problem-1/problem-1-solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

#define ROOM_WIDTH 101
#define ROOM_HEIGHT 103
#define TIME_ELAPSED 100

typedef struct {
    int x, y;   // Position
    int vx, vy; // Velocity
} Robot;

void update_position(Robot *robot, int time) {
    robot->x = (robot->x + robot->vx * time) % ROOM_WIDTH;
    if (robot->x < 0) robot->x += ROOM_WIDTH;

    robot->y = (robot->y + robot->vy * time) % ROOM_HEIGHT;
    if (robot->y < 0) robot->y += ROOM_HEIGHT;
}

void forward(Robot *robots, int count, int time, Robot *updatedRobots) {
    for (int i = 0; i < count; i++) {
        updatedRobots[i] = robots[i];
        update_position(&updatedRobots[i], time);
    }
}

void calculate_quadrants(Robot *robots, int count, int *quadrants) {
    memset(quadrants, 0, 4 * sizeof(int));
    int u = ROOM_WIDTH / 2;
    int v = ROOM_HEIGHT / 2;

    for (int i = 0; i < count; i++) {
        int x = robots[i].x;
        int y = robots[i].y;

        if (x < u && y < v) {
            quadrants[0]++;
        } else if (x > u && y < v) {
            quadrants[1]++;
        } else if (x < u && y > v) {
            quadrants[2]++;
        } else if (x > u && y > v) {
            quadrants[3]++;
        }
    }
}

int part1(Robot *robots, int count) {
    Robot updatedRobots[count];
    forward(robots, count, TIME_ELAPSED, updatedRobots);

    int quadrants[4];
    calculate_quadrants(updatedRobots, count, quadrants);

    int safetyFactor = 1;
    for (int i = 0; i < 4; i++) {
        safetyFactor *= quadrants[i];
    }
    return safetyFactor;
}

int part2(Robot *robots, int count) {
    Robot updatedRobots[count];
    bool diverged = false;

    for (int t = 0; ; t++) {
        forward(robots, count, t, updatedRobots);

        // Check uniqueness of positions
        diverged = true;
        for (int i = 0; i < count; i++) {
            for (int j = i + 1; j < count; j++) {
                if (updatedRobots[i].x == updatedRobots[j].x && updatedRobots[i].y == updatedRobots[j].y) {
                    diverged = false;
                    break;
                }
            }
            if (!diverged) break;
        }

        if (diverged) return t;
    }
}

int main() {
    FILE *file = fopen("inputdata.txt", "r");
    if (!file) {
        perror("Error opening input file");
        return EXIT_FAILURE;
    }

    char line[256];
    Robot robots[1000];
    int count = 0;

    while (fgets(line, sizeof(line), file)) {
        int x, y, vx, vy;
        sscanf(line, "p=%d,%d v=%d,%d", &x, &y, &vx, &vy);
        robots[count++] = (Robot){x, y, vx, vy};
    }

    fclose(file);

    printf("Part 1 Safety Factor: %d\n", part1(robots, count));
    printf("Part 2 Time Until All Robots Diverge: %d\n", part2(robots, count));

    return 0;
}
